### DISCUSSION TITLE
Reset last commit

### DISCUSSION THREAD

**Discussion by salimdeerhold** [community/community#74555] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

I have 3 commit i.e commit-a,commit-b,commit-c. i have git reset --soft HEAD~1
and remove commit-b and -c while preserving the changes
now the problem is that last commit is not being remove how to do that

**Comment by EliasAfara** [comment]
Could you mark my answer on your previous [discussion](https://github.com/orgs/community/discussions/74551) as the correct answer ? I would truly appreciate your help.

**Reply by salimdeerhold** [reply]
i have done u can check

**Comment by EliasAfara** [comment]
# Removing Commits with `git reset --soft`

If you've used `git reset --soft HEAD~1` to remove the last commit (`commit-c`) and preserve the changes, but the last commit (`commit-b`) is not being removed, it's likely because the soft reset only moved the `HEAD` pointer to the previous commit, effectively "uncommitting" `commit-c`. However, `commit-b` is still the tip of your branch. To remove `commit-b` as well, you can follow these steps:

## Step 1: Ensure you have no uncommitted changes

You should make sure that you don't have any uncommitted changes in your working directory. You can use `git status` to check for uncommitted changes. If there are any, you should either commit them or stash them using `git stash` to preserve them.

## Step 2: Perform an additional soft reset

To remove `commit-b` as well, you can perform another soft reset to move the `HEAD` pointer to `commit-a`. Use the following command:

```bash
git reset --soft HEAD~1
```

This command will effectively uncommit `commit-b` and move your branch back to `commit-a`.

## Step 3: Commit the changes

After you've performed the second soft reset, you'll have the changes from `commit-b` and `commit-c` staged in your working directory. You can now commit these changes to create a new commit that includes the changes from `commit-b` and `commit-c`. Use the following command:

```bash
git commit -m "New commit message"
```

Now you should have a new commit that includes the changes from `commit-b` and `commit-c`, and both `commit-b` and `commit-c` have been effectively removed from your branch's history.

**Comment by salimdeerhold** [comment]
git reset --soft HEAD~1
output:
fatal: ambiguous argument 'HEAD~1': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

**Reply by EliasAfara** [reply]
What's your current situation now ? did you manage to work through your problem ?

**Comment by salimdeerhold** [comment]
only last commit is present and i want to remove that last commit but preserve the changes

**Reply by EliasAfara** [reply]
Is your repository public? If so, I would like to take a look at it. Additionally, if you're not already using ChatGPT, I recommend using it, as it can be very helpful in finding explanations for many of your issues. I prefer not to use ChatGPT to answer your question and then copy and paste the response here

**Reply by salimdeerhold** [reply]
command -> git reset --soft HEAD~1
output->
fatal: ambiguous argument 'HEAD~1': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

i have only one commit left which was the first one.
i am doing this because the commit which was done was done by the user with worng user_name so i want to redo and commit with the corrected user_name

**Reply by salimdeerhold** [reply]
is there any other way ?

**Reply by EliasAfara** [reply]
Did you set the correct user name and email in your Git configuration ?

**Reply by salimdeerhold** [reply]
yes now cofig is correct. I need to change the commit author user name because it is showing the previous mistake user_name

**Reply by salimdeerhold** [reply]
do u know some way to do that

**Reply by EliasAfara** [reply]
Check out your commit history using `git log`, here is an example:

<img width="876" alt="image" src="https://github.com/community/community/assets/39487200/1d499aaf-e67c-44f2-b8e0-466cb030b9e4">

#

You could also make use of `git show HEAD`:

<img width="773" alt="image" src="https://github.com/community/community/assets/39487200/a7f0009a-3fd9-4d64-86d4-8599f4864548">

# 

Looking into your previous comment: you used `git reset --soft HEAD1` which is wrong. You should use `HEAD~1` (with a tilde `~`) instead of `HEAD1`. The tilde indicates that you want to reference the commit before the current `HEAD`. Here's the corrected command:

```bash
git reset --soft HEAD~1
```

**Reply by salimdeerhold** [reply]
i have used ~ sign
but while pasting in text area here it is not displaying correctly

**Comment by salimdeerhold** [comment]
git reset --soft HEAD~1
fatal: ambiguous argument 'HEAD~1': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

**Reply by EliasAfara** [reply]
send a screenshot

**Reply by salimdeerhold** [reply]
look i have again pasted with ~ sign, in fatal text also text is not displaying in prroper format

**Reply by EliasAfara** [reply]
Hmmm, is this the last commit ? Try this:

```bash
git reset --soft HEAD^
```
This command will reset the `HEAD` to the commit before the current one (the last commit), preserving the changes from that commit. The `^` symbol indicates the commit before the current one.

After running this command, you can make any necessary changes, and then create a new commit with the corrected information using `git commit`.

**Reply by salimdeerhold** [reply]
git reset --soft HEAD^ 
fatal: ambiguous argument 'HEAD^': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

and i have also done git reset --soft <commit-hash>
no error comes but also nothing happens

**Reply by EliasAfara** [reply]
run `git log` and show me by sending a screenshot.

**Reply by EliasAfara** [reply]
Let's try an alternative approach. You can use git reset with the commit hash of the previous commit to reset the HEAD.

Here's how to do it:

1. First, find the commit hash of the commit you want to reset to. You can use `git log` to view your commit history and copy the commit hash of the commit before the one you want to remove.

2. Run `git reset --soft` followed by the commit hash of the previous commit. Replace `<commit-hash>` with the actual commit hash:

```bash
git reset --soft <commit-hash>
```

For example:

```bash
git reset --soft a1b2c3d
```

**Reply by salimdeerhold** [reply]
i have tried above one too

**Comment by EliasAfara** [comment]
If this helped yo solve your issue, then that's great to know. Mark it as the correct answer even if it might not have helped you completely, others my find the solution helpful for there cases.

**Reply by salimdeerhold** [reply]
<img width="654" alt="Screenshot 2023-11-08 at 17 51 27" src="https://github.com/community/community/assets/124332008/1f0b43d6-4ac5-4840-b433-aeda2af27cdd">
when 
 git reset --soft HEAD~1

**Reply by salimdeerhold** [reply]
<img width="624" alt="Screenshot 2023-11-08 at 17 51 46" src="https://github.com/community/community/assets/124332008/921cad9d-2641-441c-acb0-57a6e0e79ddb">
git log screenshot

**Reply by EliasAfara** [reply]
Well, i am not sure what to do, you can try this is or leave it be. It's not a fatal error or a big problem.

```bash
git reset --soft <commit-hash>
```

```bash
git reset --soft 658e447bcca552e11f4bb440f39c0498e7d0f15f
```

**Comment by EliasAfara** [comment]
If you are done with this issue, close it and set my answer a the correct answer.

**Comment by airtower-luna** [comment]
> i am doing this because the commit which was done was done by the user with worng user_name so i want to redo and commit with the corrected user_name

For that you don't need to undo any commits. :slightly_smiling_face: 

To change the author of latest (or in your case, only) commit to whatever you have configured, `git commit --amend --reset-author` is enough. You could also use `--author` to explicitly set *any* author you want. See: https://git-scm.com/docs/git-commit

If you have more commits you could either do an interactive rebase and amend every commit you want to change, or use [`git-filter-repo`](https://github.com/newren/git-filter-repo). For changing Git history in general, see: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
